backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -50,30 +54,96 @@ public Match findMatch(Participant newParticipant) {
+      if (!isCombinedMatchPreferenceSatisfied(
+          combinedMatchPreference, newParticipant, currParticipant)) {
+        continue;
       }
+
+      // Found a match
+      System.out.println("found a match");
+      return new Match(
+          newParticipant.getUsername(), currParticipant.getUsername(), duration, clock.millis());
     }
     // No inital match found
     return null;
   }
+
+  /** Return true if compatible endTimeAvailable considering duration and padding */
+  private boolean isCompatibleTime(
+      int duration, long firstEndTimeAvailable, long secondEndTimeAvailable) {
+    long earliestEndTimeAvailable = Math.min(firstEndTimeAvailable, secondEndTimeAvailable);
+    return (clock.millis() + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
+        < earliestEndTimeAvailable;
+  }
+
+  /**
+   * @return combined MatchPreference similar = both are similar OR one is similar and one is any;
+   *     any = both are any; different = both are different = one is different and one is any. or
+   *     null if match preferences are not compatible
+   */
+  private MatchPreference getCombinedMatchPreference(
+      MatchPreference firstMatchPreference, MatchPreference secondMatchPreference) {
+    if ((firstMatchPreference == secondMatchPreference)
+        || (firstMatchPreference == MatchPreference.ANY)
+        || (secondMatchPreference == MatchPreference.ANY)) {
+      return (firstMatchPreference == MatchPreference.ANY)
+          ? secondMatchPreference
+          : firstMatchPreference;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * @return true if the participants are a match based on their combinedMatchPreference and the
+   *     number of fields that are the same, false if not a match
+   */
+  private boolean isCombinedMatchPreferenceSatisfied(
+      MatchPreference combinedMatchPreference,
+      Participant newParticipant,
+      Participant currParticipant) {
+    if (combinedMatchPreference != MatchPreference.ANY) {
+      // Count number of same fields between the two participants
+      boolean sameRole = newParticipant.getRole().equals(currParticipant.getRole());
+      boolean sameProductArea =
+          newParticipant.getProductArea().equals(currParticipant.getProductArea());
+      int numSameFields = Booleans.countTrue(sameRole, sameProductArea);
+      System.out.println("number of same fields: " + numSameFields);
+
+      // If match preference = similar, numSameFields must be at least MIN_SAME_FIELDS to satisfy
+      // preference
+      // If match preference = different, numSameFields must be less than MIN_SAME_FIELDS to satisfy
+      // preference
+      if ((combinedMatchPreference == MatchPreference.SIMILAR && numSameFields < MIN_SAME_FIELDS)
+          || (combinedMatchPreference == MatchPreference.DIFFERENT
+              && numSameFields >= MIN_SAME_FIELDS)) {
+        return false;
+      }
```

Comment:
Same comment about logic here. I think the two parts of "||" can be broken into two ifs, with the comments are already written.